### PR TITLE
Move outplace ops to ATen

### PR DESCRIPTION
**Pull request by zasdfgbnm**
So that things like below can be JITable, and available in C++ API:

```python
import torch

@torch.jit.script
def f(x, y, z):
    x.index_add(0, y, z)
```

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -3140,10 +3140,17 @@ def torchfn(t1, t2, t3):
...
+                    # in-place tensor is not broadcastable; test only guaranteed
+                    # to work by broadcasting other argument(s)
+                    if not hasattr(large_expanded, fn + "_"):
+                        r1 = torchfn(expanded[first], expanded[second], expanded[third])
+                        r2 = torchfn(first, second, third)
+                        self.assertEqual(r1, r2)
+                    else:
+                        r1 = torchfn(expanded[first], expanded[second], expanded[third])
+                        r2 = torchfn(expanded[first], second, third)
+                        self.assertEqual(r1, r2)
```
Comment (apaszke) [inline test/test_torch.py]: Can we just do it like this (to minimize duplication):
```python
# in-place tensor is not broadcastable; test only guaranteed
# to work by broadcasting other argument(s)
first_maybe_expand = expanded[first] if hasattr(large_expanded, fn + "_") else first
r1 = torchfn(expanded[first], expanded[second], expanded[third])
r2 = torchfn(first_maybe_expand, second, third)
self.assertEqual(r1, r2)
```

Also, it's kind of weird that those ops support this partial broadcasting. I think we should make them fully broadcasting if that's possible.

**Comment by zasdfgbnm** [comment]
@apaszke I've added support for broadcasting the first parameter.

**Comment by zasdfgbnm** [comment]
@ezyang Do you mean the autograd test for `masked_fill` and `masked_scatter`? Because the previous of outplace version of these operators was just clone and call the inplace version, which does not support broadcasting on `self`. I now add broadcast to `self`, the implementation is less trivial than a single clone, so I am just trying to test both the inplace and outpulace version separately to make sure the change does not break anything.

File: test/test_torch.py [hunk 2]
```diff
@@ -3094,59 +3094,54 @@ def _test_broadcast(self, cast):
                 # map and map2 are not implementd on CUDA tensors
                 continue
 
-            # TODO: fix masked_scatter and masked_fill broadcasting
-            if hasattr(large_expanded, fn) and fn not in ['masked_scatter', 'masked_fill']:
```
Comment (zasdfgbnm) [inline test/test_torch.py]: @ezyang this change is because `['masked_scatter', 'masked_fill']` does not support broadcasting on the first argument, so it is excluded from the test, now I need to add them back.

File: test/test_torch.py [hunk 3]
```diff
@@ -3094,59 +3094,54 @@ def _test_broadcast(self, cast):
...
+                    return myfn(t1, 1.0, t2, t3)
+                else:
+                    return myfn(t1, t2)
+
+            if hasattr(large_expanded, fn):
                 # run through tensor versions of functions
                 # and verify fully expanded inputs give same results
                 expanded = {large: large_expanded, small: small_expanded, small2: small2_expanded}
 
-                def tensorfn(myfn, t1, t2):
```
Comment (zasdfgbnm) [inline test/test_torch.py]: This is merged with the `def torchfn` below and it is now the `def specialized`

File: test/test_torch.py [hunk 4]
```diff
@@ -3094,59 +3094,54 @@ def _test_broadcast(self, cast):
...
+                    r1 = specialized(method_expanded, expanded[first], expanded[second], expanded[third])
+                    r2 = specialized(method, first, second, third)
                     self.assertEqual(r1, r2)
 
             # now for torch. versions of functions
             if hasattr(torch, fn):
                 fntorch = getattr(torch, fn)
                 expanded = {large: large_expanded, small: small_expanded, small2: small2_expanded}
 
-                def torchfn(t1, t2, t3):
```
Comment (zasdfgbnm) [inline test/test_torch.py]: This is merged with the `def tensorfn` below and it is now the `def specialized`

File: test/test_torch.py [hunk 5]
```diff
@@ -3094,59 +3094,54 @@ def _test_broadcast(self, cast):
...
+            if hasattr(large_expanded, fn):
                 # run through tensor versions of functions
                 # and verify fully expanded inputs give same results
                 expanded = {large: large_expanded, small: small_expanded, small2: small2_expanded}
 
-                def tensorfn(myfn, t1, t2):
-                    if fn == "lerp":
-                        return myfn(t1, 0.5)
-                    elif fn == "masked_select":
-                        return myfn(t1 < 0)
```
Comment (zasdfgbnm) [inline test/test_torch.py]: `masked_fill` and `masked_scatter` is not specialized here, because these two operators are explicitly excluded by the `if hasattr(large_expanded, fn) and fn not in ['masked_scatter', 'masked_fill']:` above, but now since it is no longer excluded, we need to specialize these two operators here as in the `def torchfn` below. Then this function and `def torchfn` become duplicate, so I merge them and name it `specialized`

File: test/test_autograd.py [hunk 6]
```diff
@@ -3274,21 +3274,34 @@ class dont_convert(tuple):
     ('masked_select', (), (torch.tensor(1, dtype=torch.uint8),), 'scalar'),
     ('masked_select', (M, M), (torch.tensor(1, dtype=torch.uint8),), 'scalar_broadcast_rhs'),
     ('masked_select', (), (mask_not_all_zeros((M, M)),), 'scalar_broadcast_lhs'),
+    ('masked_fill_', (M, M), (torch.ByteTensor(M, M).bernoulli_(), 10)),
```
Comment (zasdfgbnm) [inline test/test_autograd.py]: Here, since the semantics of `masked_fill` and `masked_fill_` are no longer the same (whether to support broadcast on `self`), it's better to test them separately.

File: test/test_autograd.py [hunk 7]
```diff
@@ -3274,21 +3274,34 @@ class dont_convert(tuple):
...
     ('masked_fill', (M, M), (torch.ByteTensor(M, M).bernoulli_(), 10)),
+    ('masked_fill_', (M, M), (torch.ByteTensor(M, M).bernoulli_(), torch.tensor(10)), 'tensor'),
     ('masked_fill', (M, M), (torch.ByteTensor(M, M).bernoulli_(), torch.tensor(10)), 'tensor'),
-    # no lhs or all broadcast on masked_fill or masked_scatter because it's always inplace
+    # The outplace version broadcast to lhs, the inplace version only broadcast to rhs
+    ('masked_fill', (M,), (torch.ByteTensor(M, M).bernoulli_(), 10), 'broadcast_lhs'),
+    ('masked_fill_', (M, M), (torch.ByteTensor(M,).bernoulli_(), 10), 'broadcast_rhs'),
```
Comment (apaszke) [inline test/test_autograd.py]: Isn't this already tested in the case immediately below?

File: test/test_torch.py [hunk 8]
```diff
@@ -3094,59 +3094,54 @@ def _test_broadcast(self, cast):
                 # map and map2 are not implementd on CUDA tensors
                 continue
 
-            # TODO: fix masked_scatter and masked_fill broadcasting
-            if hasattr(large_expanded, fn) and fn not in ['masked_scatter', 'masked_fill']:
+            def specialized(myfn, t1, t2, t3):
```
Comment (apaszke) [inline test/test_torch.py]: What's the meaning of specialized in here?

File: test/test_autograd.py (hunk 7 above)
Comment (zasdfgbnm) [reply test/test_autograd.py]: Hmm, yes, let me remove it.

File: test/test_torch.py [hunk 9]
```diff
@@ -3132,7 +3136,7 @@ def torchfn(t1, t2, t3):
                     elif fn == "masked_select":
                         return fntorch(t1, t2 < 0)
                     elif fn == "masked_scatter":
-                        return fntorch(t1, t2 < 0.5, cast(torch.arange(1, t1.nelement() + 1).float()))
+                        return fntorch(t1, t2 < 0.5, full1d)
```
Comment (zasdfgbnm) [inline test/test_torch.py]: Since we are adding broadcasting to the first argument, having a vector of size `t1.nelement() + 1` is no longer enough.

**Comment by zasdfgbnm** [comment]
@apaszke @ezyang I've cleaned up my changes to tests and now it should be much easier to read. The `_test_broadcast` itself is very dirty with bunch of special cases, I was trying to improve it, but seems that I made it harder to understand rather than "improve" it...

File: test/test_torch.py (hunk 9 above)
Comment (apaszke) [reply test/test_torch.py]: What do you mean by not enough?

File: aten/src/ATen/native/native_functions.yaml [hunk 10]
```diff
@@ -875,14 +875,35 @@
   # NB: This function is special-cased in tools/autograd/gen_variable_type.py
 
 - func: index_copy_(Tensor self, int64_t dim, IndexTensor index, Tensor source) -> Tensor
-  variants: method
+  variants: function, method
```
Comment (apaszke) [inline aten/src/ATen/native/native_functions.yaml]: This should not be a `function` (same applies to other in-place methods)

**Review by apaszke — COMMENTED** [review]
One minor thing (in-place method declared as a function). LGTM apart from that.

**Comment by zasdfgbnm** [comment]
@apaszke I have resolved the method vs function comment you made. The "not enough", I mean, previously `t1.masked_scatter(t2, t3)` does not support broadcasting on `t1`, therefore creating a tensor `t3` with `t1.numel()` elements large enough to make sure that it has more elements than the number of 1 masks. But now since `t1` could be expanded, `t1.numel()` elements is no longer large enough.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
@zasdfgbnm Sorry for letting this rot. It looks good to me now, could you please rebase it?

**Comment by zasdfgbnm** [comment]
@apaszke Done

**Comment by zasdfgbnm** [comment]
linter failure unrelated

**Comment by zasdfgbnm** [comment]
The failures seems unrelated and would be fixed by https://github.com/pytorch/pytorch/pull/16565

**Review by apaszke — APPROVED** [review]

**Comment by zasdfgbnm** [comment]
@apaszke I just fixed the pyi problem, now it should be ready to land.

**Comment by ezyang** [comment]
@zasdfgbnm Hey, we think this broke an (untested) codepath when people use `scatter` with a non-tensor index argument. Sample case:

```
import torch
zeros = torch.zeros(2)
zeros.scatter(0, 0, 1)
```

Do you think you could take a look? Also, it looks like we need to add an `automethod` declaration so that scatter (and the other ops) get documented.

**Comment by ezyang** [comment]
Confirmed that reverting this diff solves the problem

**Comment by zasdfgbnm** [comment]
@ezyang I would definitely take a look, but I'm pretty busy this week (I want to join PyTorch team after graduation, and busy preparing for interview :-) ), so maybe I will take a look later?

**Comment by ezyang** [comment]
OK! If you don't mind, we'll revert this for now but I'll help make sure we reland it eventually.

**Comment by zasdfgbnm** [comment]
@ezyang Feel free to revert!

**Comment by zasdfgbnm** [comment]
@ezyang Could you please reopen this for me, just for my convenience to make it an open PR on my personal https://github.com/pulls to keep track of things to do.

**Comment by gchanan** [comment]
@zasdfgbnm I think it's cleaner if you make the fix a new PR, since it will have a new number that will make it easier to differentiate if we are talking about the pre-revert state or post-revert state.

**Comment by zasdfgbnm** [comment]
@gchanan Thanks for your suggestion! I will create a new PR once I'm done with it!
